dbms/src/DataStreams/HashJoinProbeBlockInputStream.cpp

```diff
@@ -67,17 +68,109 @@ Block HashJoinProbeBlockInputStream::getHeader() const
 
 Block HashJoinProbeBlockInputStream::readImpl()
 {
-    if (probe_process_info.all_rows_joined_finish)
+    // if join finished, return {} directly.
+    if (squashing_transform.isJoinFinished())
+    {
+        return Block{};
+    }
+
+    squashing_transform.handleOverLimitBlock();
+
+    while (squashing_transform.needAppendBlock())
+    {
+        Block result_block = getOutputBlock(probe_process_info);
+        squashing_transform.appendBlock(result_block);
+    }
+    return squashing_transform.getFinalOutputBlock();
+}
+
+Block HashJoinProbeBlockInputStream::getOutputBlock(ProbeProcessInfo & probe_process_info_) const
+{
+    if (probe_process_info_.all_rows_joined_finish)
     {
         Block block = children.back()->read();
         if (!block)
+        {
             return block;
+        }
         join->checkTypes(block);
-        probe_process_info.resetBlock(std::move(block));
+        probe_process_info_.resetBlock(std::move(block));
+    }
+
+    return join->joinBlock(probe_process_info_);
+}
+
+SquashingHashJoinBlockTransform::SquashingHashJoinBlockTransform(UInt64 max_block_size_)
+    : output_rows(0)
+    , max_block_size(max_block_size_)
+    , join_finished(false)
+{}
+
+void SquashingHashJoinBlockTransform::handleOverLimitBlock()
+{
+    // we need to reset squash before handle over limit block;
+    reset();
+    // if over_limit_block is not null, we need to push it into blocks first.
+    if (over_limit_block)
+    {
+        output_rows += over_limit_block.rows();
+        blocks.push_back(std::move(over_limit_block));
+        over_limit_block = Block{};
+    }
+}
+
+void SquashingHashJoinBlockTransform::appendBlock(Block block)
+{
+    if (!block)
+    {
+        // if append block is {}, mark join finished.
+        join_finished = true;
+        return;
+    }
+    size_t current_rows = block.rows();
+
+    if (!output_rows || output_rows + current_rows <= max_block_size)
+    {
+        blocks.push_back(block);
+    }
+    else
+    {
+        // if output_rows + current_rows > max block size, put the current result block into over_limit_block and handle it in next read.
+        over_limit_block = block;
+    }
+    output_rows += current_rows;
```

Comment:
Although over_limit_block will not be output, then there is `output_rows += current_rows`?
This can be a little confusing.
We can use `return (output_rows <= max_block_size || over_limit_block) && !join_finished;`, and then don't need to `output_rows += current_rows` here.

Reply:
> Although over_limit_block will not be output, then there is `output_rows += current_rows`? This can be a little confusing. We can use `return (output_rows <= max_block_size || over_limit_block) && !join_finished;`, and then don't need to `output_rows += current_rows` here.

Add a flag over_limit to represent block over limit, only output_rows += current_rows when over_limit = false.